x-pack/plugins/lens/public/app_plugin/mounter.tsx

```diff
@@ -11,83 +11,124 @@ import { HashRouter, Route, RouteComponentProps, Switch } from 'react-router-dom
+  const embeddableEditorIncomingState = stateTransfer?.getIncomingEditorState();
+
+  const lensServices: LensAppServices = {
+    data,
+    storage,
+    navigation,
+    attributeService,
+    http: coreStart.http,
+    chrome: coreStart.chrome,
+    overlays: coreStart.overlays,
+    uiSettings: coreStart.uiSettings,
+    application: coreStart.application,
+    notifications: coreStart.notifications,
+    savedObjectsClient: coreStart.savedObjects.client,
+    getOriginatingAppName: () => {
+      return embeddableEditorIncomingState?.originatingApp
+        ? stateTransfer?.getAppNameFromId(embeddableEditorIncomingState.originatingApp)
+        : undefined;
+    },
 
+    // Temporarily required until the 'by value' paradigm is default.
+    dashboardFeatureFlag: await getByValueFeatureFlag(),
+  };
+
+  addHelpMenuToAppChrome(coreStart.chrome, coreStart.docLinks);
   coreStart.chrome.docTitle.change(
     i18n.translate('xpack.lens.pageTitle', { defaultMessage: 'Lens' })
   );
 
-  const stateTransfer = embeddable?.getStateTransfer(params.history);
-  const { originatingApp } =
-    stateTransfer?.getIncomingEditorState({ keysToRemoveAfterFetch: ['originatingApp'] }) || {};
-
-  const instance = await createEditorFrame();
-
   setReportManager(
     new LensReportManager({
-      storage: new Storage(localStorage),
       http: core.http,
+      storage,
     })
   );
-  const redirectTo = (
-    routeProps: RouteComponentProps<{ id?: string }>,
-    id?: string,
-    returnToOrigin?: boolean,
-    newlyCreated?: boolean
-  ) => {
-    if (!id) {
+
+  const getInitialInput = (
+    routeProps: RouteComponentProps<{ id?: string }>
+  ): LensEmbeddableInput | undefined => {
+    if (routeProps.match.params.id) {
+      return { savedObjectId: routeProps.match.params.id } as LensByReferenceInput;
+    }
+    if (embeddableEditorIncomingState?.valueInput) {
+      return embeddableEditorIncomingState?.valueInput as LensByValueInput;
+    }
+  };
+
+  const redirectTo = (routeProps: RouteComponentProps<{ id?: string }>, savedObjectId?: string) => {
+    if (!savedObjectId) {
       routeProps.history.push('/');
-    } else if (!originatingApp) {
-      routeProps.history.push(`/edit/${id}`);
-    } else if (!!originatingApp && id && returnToOrigin) {
-      routeProps.history.push(`/edit/${id}`);
-
-      if (newlyCreated && stateTransfer) {
-        stateTransfer.navigateToWithEmbeddablePackage(originatingApp, {
-          state: { id, type: LENS_EMBEDDABLE_TYPE },
-        });
-      } else {
-        coreStart.application.navigateToApp(originatingApp);
-      }
+    } else {
+      routeProps.history.push(`/edit/${savedObjectId}`);
     }
   };
 
+  const redirectToOrigin = (props?: RedirectToOriginProps) => {
+    if (!embeddableEditorIncomingState?.originatingApp) {
+      throw new Error('redirectToOrigin called without an originating app');
+    }
+    if (stateTransfer && props?.input) {
+      const { input, isCopied } = props;
```

Comment:
what is the use-case for `isCopied`?

Reply:
The mounter keeps track of the last `embeddableId` that was sent to it from the incomingState. When `isCopied` is false, mounter sends the same embeddableId back to dashboard, otherwise it sends 'undefined', and the dashboard creates a brand new embeddable.